---
title: "HDO.db: Human Disease Ontology Annotation Package"
author: 
    name: Erqiang Hu
    email: 13766876214@163.com
    affiliation: Southern Medical University
date: "`r Sys.Date()`"
biblio-style: apalike
output:
    prettydoc::html_pretty:
        toc: true
        theme: cayman
        highlight: github
    pdf_document:
        toc: true
vignette: >
    %\VignetteIndexEntry{HDO.db}
    %\VignettePackage{HDO.db}
    % \VignetteEngine{knitr::rmarkdown}
    % \usepackage[utf8]{inputenc}
    %\VignetteEncoding{UTF-8}
---




## Authors
Erqiang Hu

Department of Bioinformatics, School of Basic Medical Sciences, 
Southern Medical University.


## Introduction
Disease Ontology (DO) was developed to create a consistent description of gene 
products with disease perspectives, and is essential for supporting functional 
genomics in disease context. Accurate disease descriptions can discover new 
relationships between genes and disease, and new functions for previous 
uncharacteried genes and alleles.We have developed the 
[DOSE](https://bioconductor.org/packages/DOSE/)` package for semantic 
similarity analysis and disease enrichment analysis, and `DOSE` import an 
Bioconductor package 'DO.db' to get the relationship(such as parent and child)
between DO terms. But `DO.db`  hasn't been updated for years, and a lot of 
semantic information is [missing](https://github.com/YuLab-SMU/DOSE/issues/57). 
So we developed the new package `HDO.db` for Human Disease Ontology annotation.


```{r setup}
library(HDO.db)
```

## Overview

```{r}
library(AnnotationDbi)
```

The annotation data comes from 
https://github.com/DiseaseOntology/HumanDiseaseOntology/tree/main/src/ontology,
and HDO.db provide these AnnDbBimap object:
```{r}
ls("package:HDO.db")
packageVersion("HDO.db")
```

You can use `help` function to get their documents: `help(DOOFFSPRING)`
```{r}
toTable(DOmetadata)
DOMAPCOUNTS
```


## Fetch whole DO terms

In HDO.db, `DOTERM` represet the whole DO terms and their names. The users can 
also get their aliases and synonyms from `DOALIAS` and `DOSYNONYM`, 
respectively.

convert DOTERM to table
```{r}
doterm <- toTable(DOTERM)
head(doterm)
```


convert DOTERM to list
```{r}
dotermlist <- as.list(DOTERM)
head(dotermlist)
```

get alias of `DOID:0001816`
```{r}
doalias <- as.list(DOALIAS)
doalias[['DOID:0001816']]
```

get synonym of `DOID:0001816`
```{r}
dosynonym <- as.list(DOSYNONYM)
dosynonym[['DOID:0001816']]
```


## Fetch the relationship between DO terms
Similar to `DO.db`, we provide four Bimap objects to represent relationship
between DO terms: DOANCESTOR,DOPARENTS,DOOFFSPRING, and DOCHILDREN.

### DOANCESTOR
DOANCESTOR describes the association between DO terms and their ancestral terms 
based on a directed acyclic graph (DAG) defined by the Disease Ontology.
We can use `toTable` function in `AnnotationDbi` package to get a two-column 
data.frame: the first column means the DO term ids, and the second column means 
their ancestor terms.
```{r}
anc_table <- toTable(DOANCESTOR)
head(anc_table)
```

get ancestor of "DOID:0001816"
```{r}
anc_list <- AnnotationDbi::as.list(DOANCESTOR)
anc_list[["DOID:0001816"]]
```


### DOPARENTS
DOPARENTS describes the association between DO terms and their direct parent 
terms based on DAG. We can use `toTable` function in `AnnotationDbi` package to 
get a two-column data.frame: the first column means the DO term ids, and the 
second column means their parent terms.
```{r}
parent_table <- toTable(DOPARENTS)
head(parent_table)
```


get parent term of "DOID:0001816"
```{r}
parent_list <- AnnotationDbi::as.list(DOPARENTS)
parent_list[["DOID:0001816"]]
```

###  DOOFFSPRING
DOPARENTS describes the association between DO terms and their offspring  terms 
based on DAG. it's the exact opposite of `DOANCESTOR`, 
whose usage is similar to it.

get offspring of "DOID:0001816"
```{r}
off_list <- AnnotationDbi::as.list(HDO.db::DOOFFSPRING)
off_list[["DOID:0001816"]]
```

### DOCHILDREN
DOCHILDREN describes the association between DO terms and their direct children 
terms based on DAG. it's the exact opposite of `DOPARENTS`, 
whose usage is similar to it.

get children of "DOID:4"
```{r}
child_list <- AnnotationDbi::as.list(HDO.db::DOCHILDREN)
child_list[["DOID:4"]]
```

## Semantic similarity analysis
Please go to 
<https://yulab-smu.top/biomedical-knowledge-mining-book/> for the vignette.


## Disease enrichment analysis
Please go to 
<https://yulab-smu.top/biomedical-knowledge-mining-book/dose-enrichment.html> 
for the vignette.

```{r}
sessionInfo()
```
